### PR TITLE
Incorporate MuParser into fish

### PR DESCRIPTION
**Pull request by krader1961**
This is not a usable math replacement. It is the absolute bare minimum to incorporate the MuParser source and introduce a builtin `math` command. The point of this PR is to ensure no one has any objections before I take the next steps to make this actually useful per issue #3157. In particular I want people to double-check that the licensing is acceptable and that our updated license document is okay.

### CODE REVIEW

**Comment by krader1961** [comment]
Note that you can merge this and test it by typing things like `builtin math 3 + 7 / 2`. It simply isn't usable because it does not reflect the results into the parser execution context. The point of this PR is not to introduce a usable builtin `math` command. It is to verify that we can build fish with this new dependency and that the updated license is acceptable.

File: Makefile.in [hunk 1]
```diff
@@ -80,7 +89,7 @@ EXTRA_PCRE2 = @EXTRA_PCRE2@
 # Various flags
 #
 CXXFLAGS = @CXXFLAGS@ $(EXTRA_CXXFLAGS)
-CPPFLAGS = -DLOCALEDIR=\"$(localedir)\" -DPREFIX=L\"$(prefix)\" -DDATADIR=L\"$(datadir)\" -DSYSCONFDIR=L\"$(sysconfdir)\" -DBINDIR=L\"$(bindir)\" -DDOCDIR=L\"$(docdir)\" -iquote. -iquote./src/
+CPPFLAGS = -D_UNICODE=1 -DLOCALEDIR=\"$(localedir)\" -DPREFIX=L\"$(prefix)\" -DDATADIR=L\"$(datadir)\" -DSYSCONFDIR=L\"$(sysconfdir)\" -DBINDIR=L\"$(bindir)\" -DDOCDIR=L\"$(docdir)\" -iquote. -iquote./src/
```
Comment (zanchey) [inline Makefile.in]: "I've got a bad feeling about this" - I suspect `-D_UNICODE` may have side effects on Windows, although I think it's only under Visual Studio - @andrex-e-schulman may be able to comment.

File: README_muparser.md [hunk 2]
```diff
@@ -0,0 +1,3 @@
+## Math expression evaluation
```
Comment (zanchey) [inline README_muparser.md]: I don't think you need to bother with this file - perhaps put the same details into the commit messages?

File: README.md [hunk 3]
```diff
@@ -27,6 +27,7 @@ At runtime, fish requires:
 * a curses implementation such as ncurses (which should provide the `tput` command).
 * gettext, if you need the localizations.
 * PCRE2, due to the regular expression support contained in the `string` builtin. A copy is included with the source code, and will be used automatically if it does not already exist on your system.
+* MuParser for the `math` builtin. A copy is included with the source code.
```
Comment (zanchey) [inline README.md]: I'm just about to push a commit that messes this up - sorry.

File: configure.ac [hunk 4]
```diff
@@ -673,6 +677,26 @@ if test "x$included_pcre2" = "xyes"; then
   LIBS="$LIBS $PCRE2_LIBS"
 fi
 
+
+
+# ========
+# MuParser math expression library configuration.
+EXTRA_MUPARSER=
```
Comment (zanchey) [inline configure.ac]: muParser is available on some systems as an external library (eg Debian), so we'd like to be able to use the system version if it's available ala PCRE2. I can add this functionality later if you prefer.

File: configure.ac [hunk 5]
```diff
@@ -211,9 +212,10 @@ AS_IF([test "$use_doxygen" != "no"],
 AC_SYS_LARGEFILE
 
 
-# fish does not use exceptions
+# Fish does not use exceptions itself. However the MuParser library does so we
+# need to enable support for them and pay the cost of larger code. Note:
 # Disabling exceptions saves about 20% (!) of the compiled code size
-CXXFLAGS="$CXXFLAGS -fno-exceptions"
+#CXXFLAGS="$CXXFLAGS -fcxx-exceptions"
```
Comment (zanchey) [inline configure.ac]: The good news is that the comment is no longer true, at least on my system: with exceptions and debug information stripped, I get a 1.4 M fish executable; without exceptions produces a 1.3 M executable. (g++ 4.9)

File: doc_src/license.hdr [hunk 6]
```diff
@@ -9,7 +9,9 @@
 
 `fish` Copyright © 2005-2009 Axel Liljencrantz. `fish` is released under the GNU General Public License, version 2.
 
-`fish` includes other code licensed under the GNU General Public License, version 2, including GNU `printf`:
+`fish` includes other code licensed under the GNU General Public License, version 2, including GNU `printf`.
+
+`fish` includes code licensed under the MIT License, such as the MuParser math expression evaluation library.
```
Comment (zanchey) [inline doc_src/license.hdr]: The block about using the MIT license needs to go with the MIT license; this section is listing code that uses the GPL. There should be no "such as" - we should describe exactly the code that uses this license.

File: Makefile.in (hunk 1 above)
Comment (krader1961) [inline Makefile.in]: Why would we care about that? We don't support building fish as a native Windows app. And this is the only approved mechanism by the MuParser project for enabling wide character support.

File: README_muparser.md (hunk 2 above)
Comment (krader1961) [inline README_muparser.md]: It should be in the *doc_src/math.txt* file. I'll move it there.

File: configure.ac (hunk 4 above)
Comment (krader1961) [inline configure.ac]: Understood but this is just a baby step. Refinements like using the system provided library can happen later once we're confident this is our preferred solution.

File: configure.ac (hunk 5 above)
Comment (floam) [inline configure.ac]: Fish has gotten a lot bigger. :(

File: doc_src/license.hdr (hunk 6 above)
Comment (krader1961) [inline doc_src/license.hdr]: Okay, but the preceding paragraph seems to say the same thing albeit without the explicit "such as":

> `fish` includes other code licensed under the GNU General Public License, version 2, including GNU `printf`.

A normal reading of that sentence is that our `printf` implementation is only one of possibly several pieces of code licensed under GPL v2. But what do I know. IANAL :smile: So I'll drop the "such as".

**Pull request closed without merging**